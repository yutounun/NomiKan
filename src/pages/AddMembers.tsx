import React, { useState } from "react";
import { Stack } from "@mui/material";

import BaseLayout from "components/Commons/Layout/BaseLayout";
import MyHeader from "components/Commons/Organisms/MyHeader";
import MyStepper from "components/Commons/Organisms/MyStepper";
import Content from "components/AddMembers/Organisms/Content";

import MyAlert from "components/Commons/Molecules/MyAlert";

function AddMembers() {
  const [open, setOpen] = useState(false);
  const [alertLabel, setAlertLabel] = useState("");

  const handleCloseAlert = () => {
    setOpen(false);
  };

  return (
    <BaseLayout>
      <Stack direction="column" justifyContent="center" alignItems="center">
        {/* 合計金額Header */}
        <MyHeader value="メンバー追加" />

        {/* Stepper showing the current page */}
        <MyStepper activeStep={1} />

        {/* Content including input and button */}
        <Content setOpen={setOpen} setAlertLabel={setAlertLabel} />

        {/* Alert for the completion of registration  */}
        <MyAlert parentOpen={open} value={alertLabel} onClose={handleCloseAlert} />

      </Stack>
    </BaseLayout>
  );
}

export default AddMembers;
