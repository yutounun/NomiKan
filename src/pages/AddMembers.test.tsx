import { fireEvent, render, screen } from "@testing-library/react";
import React from "react";
import { BrowserRouter as Router } from "react-router-dom";
import AddMembers from "./AddMembers";

/**
 * Requirements
 * 1. The component must be rendered in the DOM xx
 * 2. TextField component must be rendered with a placeholder xx
 * 3. Reg Button component must be rendered xx
 * 4. AddButton component must be rendered xx
 * 5. If none of members is added, Add member nameしてください」 must be displayed xx
 *    and the registration button must be disabled xx
 * 6. If any characters aren't typed, the add button must be disabled xx
 * 7. TextField component has placeholder showing Add member nameしてください」 xx
 * 8. If any member is added, list the member in the member-list component xx
 * 9. The Member component has name, delete button, and edit button xx
 * 10. Clicking edit button opens Edit Member modal xx
 * 11. Clicking delete button deletes the member from the list xx
 * 12. Editing on Member modal component works
 * 13. MyStepper component is rendered with the current step xx
 * 14. Renders MyAlert component after adding a new member xx
 * 15. Shows "メンバー追加" on the title
 */

describe("AddMembers page", () => {
  it("renders correctly", () => {
    render(<Router><AddMembers /></Router>);
  });

  it("takes a snapshot", () => {
    const { asFragment } = render(<Router><AddMembers /></Router>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("has myStepper component with the current step", () => {
    render(<Router><AddMembers /></Router>);
    const AddMembersStep = screen.getByTestId("total amount");
    // Only inactive steps have "Mui-disabled" class
    expect(AddMembersStep).not.toHaveClass("Mui-disabled");

    const addMembersStep = screen.getByTestId("members");
    expect(addMembersStep).not.toHaveClass("Mui-disabled");
  });

  it("has a textField component with a placeholder", () => {
    render(<Router><AddMembers /></Router>);
    const textField = screen.getByRole("textbox", {
      name: /Add member name/i
    });
    expect(textField).toBeInTheDocument();
  });

  it("has Please add members with the disabled registration button and the disabled add button if none of members are added", () => {
    render(<Router><AddMembers /></Router>);
    const text = screen.getByText("Please add members");
    expect(text).toBeInTheDocument();

    // Make sure the registration button is disabled
    expect(screen.getByText("Submit")).toBeDisabled();
  });

  it("has a registration button component", () => {
    render(<Router><AddMembers /></Router>);
    const regBtn = screen.getByRole("button", {
      name: /Submit/i
    });
    expect(regBtn).toBeInTheDocument();
  });

  it("has a AddButton component", () => {
    render(<Router><AddMembers /></Router>);
    const addButton = screen.getByTestId("add-member-btn");
    expect(addButton).toBeInTheDocument();
  });

  it("has a Member component with name, delete button, and edit button", () => {
    render(<Router><AddMembers /></Router>);
    // Type on TextField component using fireEvent
    const textField = screen.getByRole("textbox", {
      name: /Add member name/i
    });
    fireEvent.change(textField, { target: { value: "Yuto" } });

    // Click add button
    fireEvent(
      screen.getByTestId("add-member-btn"),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    // Shows the member in the list
    const memberName = screen.getByText("Yuto");
    const deleteBtn = screen.getByTestId("delete-btn");
    const editBtn = screen.getByTestId("edit-btn");
    // Shows name in the line
    expect(memberName).toBeInTheDocument();
    // Shows delete button
    expect(deleteBtn).toBeInTheDocument();
    // Shows edit button
    expect(editBtn).toBeInTheDocument();
  });

  it("clicking delete button deletes the member from the list", () => {
    render(<Router><AddMembers /></Router>);
    // Type on TextField component using fireEvent
    const textField = screen.getByRole("textbox", {
      name: /Add member name/i
    });
    fireEvent.change(textField, { target: { value: "Yuto" } });

    // Click add button
    fireEvent(
      screen.getByTestId("add-member-btn"),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    const deleteBtn = screen.getByTestId("delete-btn");

    // Remove the member from the list
    fireEvent(
      deleteBtn,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    // The name of the added member is not displayed
    expect(screen.queryByText("Yuto")).not.toBeInTheDocument();
  });

  it("clicking edit button opens Edit Member modal with the member name", () => {
    render(<Router><AddMembers /></Router>);
    // Type on TextField component using fireEvent
    const textField = screen.getByRole("textbox", {
      name: /Add member name/i
    });
    fireEvent.change(textField, { target: { value: "Yuto" } });

    // Click add button
    fireEvent(
      screen.getByTestId("add-member-btn"),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    const editBtn = screen.getByTestId("edit-btn");

    // Open Edit Member modal
    fireEvent(
      editBtn,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(screen.getByText("EDIT")).toBeInTheDocument();
  });

  it("shows edit modal and let you register the updated name", () => {
    render(<Router><AddMembers /></Router>);
    // Type on TextField component using fireEvent
    const textField = screen.getByRole("textbox", {
      name: /Add member name/i
    });
    fireEvent.change(textField, { target: { value: "Yuto" } });

    // Click add button
    fireEvent(
      screen.getByTestId("add-member-btn"),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    const editBtn = screen.getByTestId("edit-btn");

    // Open Edit Member modal
    fireEvent(
      editBtn,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(screen.getByText("Yuto")).toBeInTheDocument();

    const textEditField = screen.getByRole("textbox");
    fireEvent.change(textEditField, { target: { value: "Taka" } });

    const submitEditBtn = screen.getByTestId("submit-edit-btn");

    // Close Edit Member modal
    fireEvent(
      submitEditBtn,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(screen.getByText("Taka")).toBeInTheDocument();
  });

  it("close edit modal without updating", () => {
    render(<Router><AddMembers /></Router>);
    // Type on TextField component using fireEvent
    const textField = screen.getByRole("textbox", {
      name: /Add member name/i
    });
    fireEvent.change(textField, { target: { value: "Yuto" } });

    // Click add button
    fireEvent(
      screen.getByTestId("add-member-btn"),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    const editBtn = screen.getByTestId("edit-btn");

    // Open Edit Member modal
    fireEvent(
      editBtn,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    const textEditField = screen.getByRole("textbox");
    fireEvent.change(textEditField, { target: { value: "Taka" } });
    const closeBtn = screen.getByTestId("cancel-edit-btn");

    // Close Edit Member modal
    fireEvent(
      closeBtn,
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(screen.getByText("Yuto")).toBeInTheDocument();
  });

  it("does not render alert at first", () => {
    render(<Router><AddMembers /></Router>);
    expect(screen.queryByRole("alert")).toBeFalsy();
  });

  it("renders MyAlert component after adding a new member", () => {
    render(<Router><AddMembers /></Router>);
    // Type on TextField component using fireEvent
    const textField = screen.getByRole("textbox", {
      name: /Add member name/i
    });
    fireEvent.change(textField, { target: { value: "Yuto" } });

    // Click add button
    fireEvent(
      screen.getByTestId("add-member-btn"),
      new MouseEvent("click", {
        bubbles: true,
        cancelable: true,
      })
    );

    expect(screen.getByText(/Add new member!!/i)).toBeInTheDocument();
  });

  it("Shows 'Members' on the title", () => {
    render(<Router><AddMembers /></Router>);
    const header = screen.getByRole("heading", {
      name: /Members/i,
    });

    expect(header).toBeInTheDocument();
  });
});
