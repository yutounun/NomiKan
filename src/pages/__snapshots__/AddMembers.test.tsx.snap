// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddMembers page takes a snapshot 1`] = `
<DocumentFragment>
  <main
    data-testid="main"
  >
    <div
      class="MuiStack-root css-w004l2-MuiStack-root"
    >
      <div
        class="MuiStack-root css-wzdyko-MuiStack-root"
        data-testid="header"
      >
        <h1
          class="MuiTypography-root MuiTypography-h1 css-nqt6fq-MuiTypography-root"
        >
          Members
        </h1>
      </div>
      <div
        class="MuiStack-root css-9922yn-MuiStack-root"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="ArrowBackIosIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.67 3.87 9.9 2.1 0 12l9.9 9.9 1.77-1.77L3.54 12z"
          />
        </svg>
      </div>
      <div
        class="MuiBox-root css-jxhp8u"
      >
        <div
          class="MuiStepper-root MuiStepper-horizontal MuiStepper-alternativeLabel css-10mg1vw-MuiStepper-root"
          data-testid="stepper"
        >
          <div
            class="MuiStep-root MuiStep-horizontal MuiStep-alternativeLabel Mui-completed css-n7tliy-MuiStep-root"
          >
            <span
              class="MuiStepLabel-root MuiStepLabel-horizontal MuiStepLabel-alternativeLabel css-ascpo7-MuiStepLabel-root"
              data-testid="total amount"
            >
              <span
                class="MuiStepLabel-iconContainer Mui-completed MuiStepLabel-alternativeLabel css-vnkopk-MuiStepLabel-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root Mui-completed css-1u4zpwo-MuiSvgIcon-root-MuiStepIcon-root"
                  data-testid="CheckCircleIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 0a12 12 0 1 0 0 24 12 12 0 0 0 0-24zm-2 17l-5-5 1.4-1.4 3.6 3.6 7.6-7.6L19 8l-9 9z"
                  />
                </svg>
              </span>
              <span
                class="MuiStepLabel-labelContainer MuiStepLabel-alternativeLabel css-1vyamtt-MuiStepLabel-labelContainer"
              >
                <span
                  class="MuiStepLabel-label Mui-completed MuiStepLabel-alternativeLabel css-1hv8oq8-MuiStepLabel-label"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                  >
                    total amount
                  </p>
                </span>
              </span>
            </span>
          </div>
          <div
            class="MuiStep-root MuiStep-horizontal MuiStep-alternativeLabel css-n7tliy-MuiStep-root"
          >
            <div
              class="MuiStepConnector-root MuiStepConnector-horizontal MuiStepConnector-alternativeLabel Mui-active css-zpcwqm-MuiStepConnector-root"
            >
              <span
                class="MuiStepConnector-line MuiStepConnector-lineHorizontal css-z7uhs0-MuiStepConnector-line"
              />
            </div>
            <span
              class="MuiStepLabel-root MuiStepLabel-horizontal MuiStepLabel-alternativeLabel css-ascpo7-MuiStepLabel-root"
              data-testid="members"
            >
              <span
                class="MuiStepLabel-iconContainer Mui-active MuiStepLabel-alternativeLabel css-vnkopk-MuiStepLabel-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root Mui-active css-1u4zpwo-MuiSvgIcon-root-MuiStepIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="12"
                  />
                  <text
                    class="MuiStepIcon-text css-117w1su-MuiStepIcon-text"
                    dominant-baseline="central"
                    text-anchor="middle"
                    x="12"
                    y="12"
                  >
                    2
                  </text>
                </svg>
              </span>
              <span
                class="MuiStepLabel-labelContainer MuiStepLabel-alternativeLabel css-1vyamtt-MuiStepLabel-labelContainer"
              >
                <span
                  class="MuiStepLabel-label Mui-active MuiStepLabel-alternativeLabel css-1hv8oq8-MuiStepLabel-label"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                  >
                    members
                  </p>
                </span>
              </span>
            </span>
          </div>
          <div
            class="MuiStep-root MuiStep-horizontal MuiStep-alternativeLabel css-n7tliy-MuiStep-root"
          >
            <div
              class="MuiStepConnector-root MuiStepConnector-horizontal MuiStepConnector-alternativeLabel Mui-disabled css-zpcwqm-MuiStepConnector-root"
            >
              <span
                class="MuiStepConnector-line MuiStepConnector-lineHorizontal css-z7uhs0-MuiStepConnector-line"
              />
            </div>
            <span
              class="MuiStepLabel-root MuiStepLabel-horizontal Mui-disabled MuiStepLabel-alternativeLabel css-ascpo7-MuiStepLabel-root"
              data-testid="ratio"
            >
              <span
                class="MuiStepLabel-iconContainer Mui-disabled MuiStepLabel-alternativeLabel css-vnkopk-MuiStepLabel-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root css-1u4zpwo-MuiSvgIcon-root-MuiStepIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="12"
                  />
                  <text
                    class="MuiStepIcon-text css-117w1su-MuiStepIcon-text"
                    dominant-baseline="central"
                    text-anchor="middle"
                    x="12"
                    y="12"
                  >
                    3
                  </text>
                </svg>
              </span>
              <span
                class="MuiStepLabel-labelContainer MuiStepLabel-alternativeLabel css-1vyamtt-MuiStepLabel-labelContainer"
              >
                <span
                  class="MuiStepLabel-label Mui-disabled MuiStepLabel-alternativeLabel css-1hv8oq8-MuiStepLabel-label"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                  >
                    ratio
                  </p>
                </span>
              </span>
            </span>
          </div>
          <div
            class="MuiStep-root MuiStep-horizontal MuiStep-alternativeLabel css-n7tliy-MuiStep-root"
          >
            <div
              class="MuiStepConnector-root MuiStepConnector-horizontal MuiStepConnector-alternativeLabel Mui-disabled css-zpcwqm-MuiStepConnector-root"
            >
              <span
                class="MuiStepConnector-line MuiStepConnector-lineHorizontal css-z7uhs0-MuiStepConnector-line"
              />
            </div>
            <span
              class="MuiStepLabel-root MuiStepLabel-horizontal Mui-disabled MuiStepLabel-alternativeLabel css-ascpo7-MuiStepLabel-root"
              data-testid="result"
            >
              <span
                class="MuiStepLabel-iconContainer Mui-disabled MuiStepLabel-alternativeLabel css-vnkopk-MuiStepLabel-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiStepIcon-root css-1u4zpwo-MuiSvgIcon-root-MuiStepIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="12"
                  />
                  <text
                    class="MuiStepIcon-text css-117w1su-MuiStepIcon-text"
                    dominant-baseline="central"
                    text-anchor="middle"
                    x="12"
                    y="12"
                  >
                    4
                  </text>
                </svg>
              </span>
              <span
                class="MuiStepLabel-labelContainer MuiStepLabel-alternativeLabel css-1vyamtt-MuiStepLabel-labelContainer"
              >
                <span
                  class="MuiStepLabel-label Mui-disabled MuiStepLabel-alternativeLabel css-1hv8oq8-MuiStepLabel-label"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 css-e784if-MuiTypography-root"
                  >
                    result
                  </p>
                </span>
              </span>
            </span>
          </div>
        </div>
      </div>
      <div
        class="MuiStack-root css-1qvmp2n-MuiStack-root"
      >
        <div
          class="MuiStack-root css-7owdng-MuiStack-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1xa5aqc-MuiFormControl-root-MuiTextField-root"
          >
            <label
              class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeSmall MuiInputLabel-outlined css-1pysi21-MuiFormLabel-root-MuiInputLabel-root"
              data-shrink="false"
              for=":r1:"
              id=":r1:-label"
            >
              Add member name
            </label>
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <input
                aria-invalid="false"
                class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
                id=":r1:"
                type="string"
                value=""
              />
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-yjsfm1"
                >
                  <span>
                    Add member name
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
          <button
            aria-label="delete"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
              data-testid="add-member-btn"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M13 7h-2v4H7v2h4v4h2v-4h4v-2h-4V7zm-1-5C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiStack-root css-hmog4a-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-1wn5ih4-MuiTypography-root"
          >
            Please add members
          </p>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1fs9aek-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <span
            class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="DoneOutlineIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m19.77 5.03 1.4 1.4L8.43 19.17l-5.6-5.6 1.4-1.4 4.2 4.2L19.77 5.03m0-2.83L8.43 13.54l-4.2-4.2L0 13.57 8.43 22 24 6.43 19.77 2.2z"
              />
            </svg>
          </span>
          Submit
        </button>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
